### DISCUSSION TITLE
updateOne & updateMany Response Mapping

### DISCUSSION THREAD

**Discussion by kathirsvn** [stargate/stargate-mongoose#24] [General]
@vkarpov15  @amorton @maheshrajamani 

The response of `updateOne` command from the JSON API needs to be mapped to the following fields of mongoose updateOne API response
```
acknowledged
modifiedCount
matchedCount
upsertedCount
upsertedId
```

`acknowledged` is always true since our API operations are synchronous and there is no fire and forget kind of requests

For one of the tests of `updateOne` operation (without 'upsert' flag), we are getting below response.
`{"status":{"updatedIds":["7505ffec-47a0-4346-8ee3-72fd70ad09e5"]}}`

**Question 1:** Does this look good if I map it as below?

```
modifiedCount: 1 (i.e. status.updatedIds?.length)
matchedCount: 1 (i.e. status.updatedIds?.length)
upsertedCount: 0
upsertedId: 0
```

_If more than one record is matched, but only one record is updated, the matchedCount will still be 1?_

**Question 2:** When there is 'upsert' flag in the request and if an upsert happened, I assume we would be getting below response.
`{"status":{"upsertedId":"7505ffec-47a0-4346-8ee3-72fd70ad09e5"}}`
So, in that case, is it good if I map the result as below?

```
modifiedCount: 0
matchedCount: 0
upsertedCount: 1 (status.upsertedId ? 1 : 0)
upsertedId: "7505ffec-47a0-4346-8ee3-72fd70ad09e5" (status.upsertedId)
```

**Question 3:** Also, `updateMany` has same signature. Is the same mapping logic applicable for `updateMany` as well?

**Comment by amorton** [comment]
> For one of the tests of updateOne operation (without 'upsert' flag), we are getting below response.
{"status":{"updatedIds":["7505ffec-47a0-4346-8ee3-72fd70ad09e5"]}}

This is a bug on the server side, it should return the matchedCount etc. 

Can you raise a bug https://github.com/stargate/jsonapi that updateOne (and other commands that support upsert) should return the same fields as when upsert is false. 

when we do an upsert there should be a field in the status object called `upsertedId` that is the value of the new ID. It should not be an array. 

(the spec is being here) 

 > Question 1: Does this look good if I map it as below?

Neither, we should be getting the data we need from the server. 

> Question 2: When there is 'upsert' flag in the request and if an upsert happened, I assume we would be getting below response.
{"status":{"upsertedId":"7505ffec-47a0-4346-8ee3-72fd70ad09e5"}}
So, in that case, is it good if I map the result as below?

We should return a map of these values, all returned from the server other than acknowledged: 

```
acknowledged: true
matchedCount: : 
modifiedCount:  
upsertedId: 
```

> Question 3: Also, updateMany has same signature. Is the same mapping logic applicable for updateMany as well?
Yes, and the server should be returning the correct data as well :) 

Again spec is behind, we need to get in sync.

**Comment by vkarpov15** [comment]
1. That's fine. My only suggestion would be to make `upsertedId` undefined rather than 0. "If more than one record is matched, but only one record is updated, the matchedCount will still be 1?" <-- correct for `updateOne`.
2. Correct. In particular, `matchedCount` and `modifiedCount` should both be 0. `modifiedCount` returns the number of _existing_ documents updated, upserting a new document doesn't count.

**Reply by kathirsvn** [reply]
Thanks @vkarpov15 !
